src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodFilteringService.cs

```diff
@@ -0,0 +1,191 @@
+                project.Solution,
+                nameof(IRemoteExtensionMethodFilteringService.GetPossibleExtensionMethodMatchesAsync),
+                new object[] { project.Id, targetTypeNames.ToArray(), loadOnly },
+                cancellationToken).ConfigureAwait(false);
+
+            if (!remoteResults.Any())
+            {
+                return null;
+            }
+
+            // Reconstruct the dictionary from remote results.
+            var results = new MultiDictionary<string, string>();
+
+            foreach (var pair in remoteResults)
+            {
+                foreach (var value in pair.Item2)
+                {
+                    results.Add(pair.Item1, value);
+                }
+            }
+
+            return results;
+        }
+
+        private static async Task<MultiDictionary<string, string>?> GetPossibleExtensionMethodMatchesInCurrentProcessAsync(Project currentProject, ImmutableHashSet<string> targetTypeNames, bool loadOnly, CancellationToken cancellationToken)
+        {
+            var solution = currentProject.Solution;
+            var graph = currentProject.Solution.GetProjectDependencyGraph();
+            var relevantProjectIds = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id)
+                                        .Concat(currentProject.Id);
+
+            using var syntaxDisposer = ArrayBuilder<SyntaxTreeIndex>.GetInstance(out var syntaxTreeIndexBuilder);
+            using var symbolDisposer = ArrayBuilder<SymbolTreeInfo>.GetInstance(out var symbolTreeInfoBuilder);
+            var peReferences = PooledHashSet<PortableExecutableReference>.GetInstance();
+
+            try
+            {
+                foreach (var projectId in relevantProjectIds.Concat(currentProject.Id))
+                {
+                    // Alway create indices for documents in current project if they don't exist.
+                    loadOnly &= projectId != currentProject.Id;
+
+                    var project = solution.GetProject(projectId);
+                    if (project == null || !project.SupportsCompilation)
+                    {
+                        continue;
+                    }
+
+                    // Transitively get all the PE references
+                    peReferences.AddRange(project.MetadataReferences.OfType<PortableExecutableReference>());
+                    foreach (var document in project.Documents)
+                    {
+                        // Don't look for extension methods in generated code.
+                        if (document.State.Attributes.IsGenerated)
+                        {
+                            continue;
+                        }
+
+                        var info = await document.GetSyntaxTreeIndexAsync(loadOnly, cancellationToken).ConfigureAwait(false);
+
+                        // Don't provide anyting if we don't have all the required SyntaxTreeIndex created.
```

Comment:
1. anyting
2. this seems like a broad brush.  why not provide from what we have created indices for?

Reply:
Since the filter is implemented as a whitelist, so it won't work unless we have all the data.

Reply:
> Since the filter is implemented as a whitelist, so it won't work unless we have all the data.

I don't think i understand what this means.

Reply:
What I meant is, the "filter" would return a superset of final results. So if you use a partial filter, we might miss things from the missing indices.

Reply:
But as you mentioned before, intellisense doesn't need to be complete, so it might be OK to do so, people just need to understand that in extreme cases a lot of thing might be missing. Given that we now try to populate the indices in the background once we hit this, next time they will likely be available.

Reply:
For now, I'm inclined to still do the quick bail, i.e. not return partial results.